include/RAJA/util/zip_tuple.hpp

```diff
@@ -0,0 +1,348 @@
+
+/*!
+    \brief Call f on each member of t (f(t)...).
+*/
+template < typename Tuple, typename F >
+RAJA_HOST_DEVICE inline
+void zip_for_each(Tuple&& t, F&& f)
+{
+  zip_for_each_impl(std::forward<Tuple>(t), std::forward<F>(f), typename camp::decay<Tuple>::IdxSeq{});
+}
+
+/*!
+    \brief Call f on each member of t0 and t1 (f(t0, t1)...).
+*/
+template < typename Tuple0, typename Tuple1, typename F >
+RAJA_HOST_DEVICE inline
+void zip_for_each(Tuple0&& t0, Tuple1&& t1, F&& f)
+{
+  static_assert(std::is_same<typename camp::decay<Tuple0>::IdxSeq, typename camp::decay<Tuple1>::IdxSeq>::value,
+      "Tuple0 and Tuple1 must have the same size");
+  zip_for_each_impl(std::forward<Tuple0>(t0), std::forward<Tuple1>(t1), std::forward<F>(f), typename camp::decay<Tuple0>::IdxSeq{});
+}
+
+
+/*!
+    \brief Tuple used by ZipIterator for storing multiple references and values.
+    Acts like a reference to its members allowing copy/move construction/assignment
+    based on the reference type of the zip_tuple.
+*/
+template < bool is_val, typename ... Ts >
+struct zip_tuple
+{
+  using value_type = RAJA::tuple<Ts...>;
+
+  // zip_tuple type with opposite is_val
+  template < typename ... Os >
+  using opp_tuple = zip_tuple<!is_val, Os...>;
+
+  // camp::idx_seq for this type, also used by zip_for_each
+  using IdxSeq = camp::make_idx_seq_t<sizeof...(Ts)>;
+
+  // constructor from types convertible to Ts
+  template < typename ... Os
+           , typename = concepts::enable_if<type_traits::convertible_to<Os&&, Ts>...> >
+  RAJA_HOST_DEVICE RAJA_INLINE zip_tuple(Os&&... os)
+    : m_tuple(std::forward<Os>(os)...) { }
+
+  // assignment from types convertible to Ts
+  template < typename ... Os
+           , typename = concepts::enable_if<type_traits::convertible_to<Os&&, typename std::remove_reference<Ts>::type>...> >
+  zip_tuple& assign(Os&&... os)
+  { return assign_helper(IdxSeq{}, std::forward<Os>(os)...); }
+
+  // copy and move constructors
+  RAJA_HOST_DEVICE RAJA_INLINE zip_tuple(zip_tuple &      o)
+    : zip_tuple(          o , IdxSeq{}) { }
+  RAJA_HOST_DEVICE RAJA_INLINE zip_tuple(zip_tuple const& o)
+    : zip_tuple(          o , IdxSeq{}) { }
+  RAJA_HOST_DEVICE RAJA_INLINE zip_tuple(zip_tuple &&     o)
+    : zip_tuple(std::move(o), IdxSeq{}) { } // move if is_val, pass-through otherwise
+
+  // copy and move assignment operators
+  RAJA_HOST_DEVICE RAJA_INLINE zip_tuple& operator=(zip_tuple &      o)
+  { return assign_helper(          o , IdxSeq{}); }
+  RAJA_HOST_DEVICE RAJA_INLINE zip_tuple& operator=(zip_tuple const& o)
+  { return assign_helper(          o , IdxSeq{}); }
+  RAJA_HOST_DEVICE RAJA_INLINE zip_tuple& operator=(zip_tuple &&     o)
+  { return assign_helper(std::move(o), IdxSeq{}); }
+
+  // copy and move constructors from opp_tuple type zip_tuples
+  template < typename ... Os, typename = typename std::enable_if<sizeof...(Ts) == sizeof...(Os)>::type >
+  RAJA_HOST_DEVICE RAJA_INLINE zip_tuple(opp_tuple<Os...> &      o)
```

Comment:
Remove Os... pack and replace with conditional (based on is_val) add_reference<Ts>::type